flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/ExprCodeGenerator.scala

```diff
@@ -489,7 +489,7 @@ class ExprCodeGenerator(ctx: CodeGeneratorContext, nullableInput: Boolean)
       // We only support JSON function operands as the value param of a JSON_OBJECT or JSON_ARRAY function
       case (operand: RexNode, i)
           if isJsonFunctionOperand(operand) &&
-            (isJsonArrayOperand(call) || i == 2 && isJsonObjectOperand(call)) =>
+            (isJsonArrayOperand(call) || isJsonObjectOperand(call) && inValuesParam(i)) =>
```

Comment:
i wonder whether we can extract the whole `if` condition into method

Reply:
Created a helper function called `isSupportedJsonOperand`